0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1181 @@
+
+**Metadata URL Name:** `transfer_fulfillment`. Suggested generic path: `/transfers/{client_id}/fulfillment`
+
+**Caution:** The ledger _MUST_ require the `Content-Type: application/json` header on this request. In the future, the API might also accept fulfillments in binary fulfillment at this same endpoint.
+
+##### Body Parameters
+
+The message body should be a JSON object with one field, `fulfillment`, which contains a [Crypto-Condition Fulfillment][] in string format.
+
+#### Response Format
+
+A successful result uses the HTTP response code **201 Created**. The message body of the response is a JSON object containing the submitted fulfillment in the `fulfillment` field. A ledger MUST return a successful response if and only if the transfer was executed as a result of this request.
+
+If you
+
+#### Example
+
+Request:
+
+```json
+PUT /transfers/3a2a1d9e-8640-4d2d-b06c-84f2cd613204/fulfillment
+Content-Type: application/json
+
+{
+  "fulfillment": "cf:0:_v8"
+}
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+
+{
+  "fulfillment": "cf:0:_v8"
+}
+```
+
+#### Errors
+
+- [NotFoundError][] - The transfer does not exist.
+- [UnmetConditionError][] - The fulfillment does not match the condition.
+- [TransferNotConditionalError][] - The transfer had no condition to fulfill.
+- [TransferStateError][] - The transfer was not in the `prepared` state when the request was received. This occurs if the transfer has already been executed, rejected, or expired.
+- [UnprocessableEntityError][] - The request is formatted properly but contains an otherwise-unspecified semantic problem.
+- [InvalidBodyError][] - The request body was not properly-formatted JSON.
+
+
+### Reject Transfer
+[Reject Transfer]: #reject-transfer
+
+Reject a prepared transfer. A transfer can be rejected if and only if that transfer is in the `prepared` state. Doing so transitions the transfer to the `rejected` state.
+
+**Authorization:** The `credit_account` of a transfer MUST be able to reject the transfer. Unauthorized users and the `debit_account` MUST NOT be able to reject the transfer. Other account owners MUST NOT be able to reject the transfer unless those accounts are explicitly authorized to do so by the `credit_account` and those accounts are not the `debit_account`.
+
+#### Request Format
+
+PUT to the `transfer_rejection` path of the transfer you want to execute, with a JSON object in the message body containing a rejection reason.
+
+**Metadata URL Name:** `transfer_rejection`. Suggested generic path: `/transfers/{client_id}/rejection`
+
+##### Body Parameters
+
+The message body should be a JSON object with one top-level field:
+
+| Field              | Value  | Description                                    |
+|:-------------------|:-------|:-----------------------------------------------|
+| `rejection_reason` | String | The reason for rejecting the transfer. This is intended to be a machine-readable identifier. |
```

Comment:
expired, unwanted, insufficient_liquidity or insufficient_time (for the connector to pass it on), invalid_rate, unsupported_condition...

Now that I'm thinking about this though I'm wondering if it actually makes sense to have this field in the transfer at all.

We need a protocol for communicating errors down the path analogous to ICMP. That will be defined on the same level as the quoting protocol, and would probably use the message functionality for communication. Since we need that protocol anyway, I'm not sure it makes sense to either define `rejection_reason`s on this level or even include them on this level. Most of the reasons will be _inter_-ledger ones anyway, so the messages used should be more separate from this ledger layer.